Fix regression in wxDC::Clear() and make wxGCDC::Clear() consistent with it

vadz opened the pull request:
This fixes [#18371](http://trac.wxwidgets.org/ticket/18371) after the changes done to fix [#10273](http://trac.wxwidgets.org/ticket/10273) broke it by making `Clear()` a NOP unless the background brush had been explicitly set, instead of using the white brush by default as was done previously.

vadz committed:
Clear DC using white background by default in wxMSW

Since 848f5e78a60bde49900159a77b218bb15c80382d the background wasn't
cleared at all if the background brush hadn't been explicitly set prior
to calling wxDC::Clear(). This was incompatible with the old behaviour
and even managed to break our own print preview code, so it clearly
wasn't a good idea to change Clear() like this.

Instead, continue to clear the DC using white background by default,
while still not doing anything if a transparent background brush had
been explicitly set.

This fixes print preview background under MSW.

See #10273, #18371.

vadz committed:
Remove useless code in wxGCDCImpl::SetBackground()

No changes.

vadz committed:
Revert "Use white background when rendering print preview"

This reverts commit 70f0900799b670107982d683cecf216453fc67fe.

It is not necessary any more after fixes to wxDC::Clear() in the
previous commits.

vadz committed:
Clear background if background brush is invalid in wxGTK2 too

We only should avoid clearing the background if the brush is
transparent, not if it's not set.

This refines 6549d4c3c5f5d3e3acc754a2b1dc04f4a0a336d1

See #10273.

vadz committed:
Don't clear background with transparent brush in wxGCDC

Bail out immediately if the background brush is transparent, for
consistency with the other wxDC implementations.

See #10273.

vadz committed:
Clear background with white brush by default in wxGCDC

This makes wxGCDC consistent with wxDC.

This commit is best viewed ignoring whitespace.

paulcor reviewed (commented):
The documentation for `wxDC::Clear()` doesn't say anything about using white by default, and the behavior wasn't consistent across platforms, so I don't agree that this is a regression. If you really want to make white the default, and rely on that in our own code, it should probably be documented that way.

vadz commented on src/gtk/dcclient.cpp:
```diff
@@ -1508,7 +1508,7 @@ void wxWindowDCImpl::Clear()
     if (!m_gdkwindow) return;
 
-    if (!m_backgroundBrush.IsOk() || m_backgroundBrush.GetStyle() == wxBRUSHSTYLE_TRANSPARENT)
+    if (m_backgroundBrush.IsTransparent())
```
OK, I'll fix this.

vadz commented:
> The documentation for `wxDC::Clear()` doesn't say anything about using white by default, and the behavior wasn't consistent across platforms, so I don't agree that this is a regression.

It was de factor a regression because it broke print preview under both MSW and GTK (and probably elsewhere too). And if we relied on `Clear()` using white by default in our own code, I'm willing to bet that there is code outside of wx relying on the same thing.

Also, it was consistent between MSW and GTK2, originally, and clearly the latter used `wxWHITE_BRUSH` intentionally, so I don't think there is any real doubt about the wanted behaviour here, is there?

> If you really want to make white the default, and rely on that in our own code, it should probably be documented that way.

Yes, indeed, I should have documented it, thanks for the reminder. I'll do it now.

vadz committed:
Document that wxDC::Clear() uses white background by default

This officially documents the historical behaviour of this method in
wxMSW and wxGTK2.

vadz committed:
Add unit tests for wxDC::Clear()

Verify that it behaves as documented, i.e. uses white if the background
brush is not explicitly set.

vadz committed:
Use white in wxDC::Clear() if background brush is reset in wxGTK2

Although white background was used if the brush was never set, setting
and resetting it nothing unexpectedly (and inconsistently with the other
ports) resulted in using black background.

Fix this by just resetting the brush to white if it's invalid, instead
of not setting it at all.

vadz commented:
@paulcor Did you have a chance to look at the updated version by chance?
